### PR TITLE
Surface GPU support in qsimcirq

### PR DESCRIPTION
**Pull request by 95-martin-orion**
This PR will allow qsimcirq users to request simulation on GPU resources.

### CODE REVIEW

**Comment by 95-martin-orion** [comment]
This PR is now ready for review. Main components include:
* Added CUDA pybind module
* Modified existing pybind to work with both CPU and GPU
* Updated Makefile to build CUDA module if NVCC is available
* Added `QSimOptions` to contain Python simulator options
* Testing for GPU (verified on system w/GPU; skipped if unavailable)

Users with an NVIDIA GPU and the CUDA toolkit can make use of this feature by:
* Compiling qsim locally to enable GPU support
* Creating a qsimcirq.Simulator with `QSimOptions(use_gpu=True)`
* Executing circuits with that simulator

Known issues:
* Simulating expectation values can fail if the observables are composed of 1-qubit gates; this will cause the GPU expectation value test to fail. Observables on 2+ qubits should be unaffected.

**Comment by 95-martin-orion** [comment]
The "known issue" was resolved in #409.

**Review by sergeisakov — CHANGES_REQUESTED** [review]
Thank you! It looks good. There are just a few small issues.

**Comment by sergeisakov** [comment]
> Known issues:
> 
> * Simulating expectation values can fail if the observables are composed of 1-qubit gates; this will cause the GPU expectation value test to fail. Observables on 2+ qubits should be unaffected.

Could you remove that?

**Comment by sergeisakov** [comment]
qsimh_simulator.py is not updated to run on GPUs, right?

**Review by 95-martin-orion — COMMENTED** [review]
Addressed the quick fixes. Still need to verify on GPU and fix+test the `init_state` concern.

**Comment by 95-martin-orion** [comment]
> qsimh_simulator.py is not updated to run on GPUs, right?

Correct, it remains CPU-only.

File: pybind_interface/pybind_main.cpp [hunk 1]
```diff
@@ -573,7 +590,7 @@ class SimulatorHelper {
                 float* fsv) {
       fsv[i] = ptr[i];
     };
-    For(num_threads).Run(input_vector.size(), f, ptr, state.get());
+    For(num_sim_threads).Run(input_vector.size(), f, ptr, state.get());
```
[the remark this answers is not in the document]
Comment (95-martin-orion) [reply pybind_interface/pybind_main.cpp]: Done and done.

**Review by sergeisakov — APPROVED** [review]

**Pull request merged**